File: app/src/main/java/me/pwcong/jpstart/component/activity/BaseActivity.java
package me.pwcong.jpstart.component.activity;

import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.view.View;

import butterknife.ButterKnife;
import me.pwcong.jpstart.R;
import me.pwcong.jpstart.manager.ActivityManager;

/**
 * Created by Pwcong on 2016/9/23.
 */

public abstract class BaseActivity extends AppCompatActivity {


    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        ActivityManager.getInstance().register(this);
        ActivityManager.setCurrent(this);

        setContentView(getViewId());

        ButterKnife.bind(this);

        initVariable(savedInstanceState);

        doAction();

    }

    protected abstract int getViewId();

    protected abstract void initVariable(@Nullable Bundle savedInstanceState);

    protected abstract void doAction();

    @Override
    protected void onDestroy() {
        super.onDestroy();

        ActivityManager.getInstance().unregister(this);

    }

    public void showSnackBar(View view, String msg) {
        Snackbar.make(view, msg, Snackbar.LENGTH_SHORT).show();
    }

    public void showSnackBar(View view, int msg) {
        Snackbar.make(view, msg, Snackbar.LENGTH_SHORT).show();
    }

    public void showAlertDialog(Context context, int titleId, String message,
                                int iconId,
                                int positiveTextId, DialogInterface.OnClickListener positiveButtonListener) {

        new AlertDialog.Builder(context)
                .setTitle(titleId)
                .setMessage(message)
                .setPositiveButton(positiveTextId, positiveButtonListener)
                .setIcon(iconId)
                .create()
                .show();

    }

    public void showAlertDialog(Context context, int titleId, int messageId,
                                int iconId,
                                int positiveTextId, DialogInterface.OnClickListener positiveButtonListener) {

        new AlertDialog.Builder(context)
                .setTitle(titleId)
                .setMessage(messageId)
                .setPositiveButton(positiveTextId, positiveButtonListener)
                .setIcon(iconId)
                .create()
                .show();

    }

    public void showAlertDialog(Context context, int titleId, String message,
                                int iconId,
                                int positiveTextId, DialogInterface.OnClickListener positiveButtonListener,
                                int negativeTextId, DialogInterface.OnClickListener negativeButtonListener) {

        new AlertDialog.Builder(context)
                .setTitle(titleId)
                .setMessage(message)
                .setPositiveButton(positiveTextId, positiveButtonListener)
                .setNegativeButton(negativeTextId, negativeButtonListener)
                .setIcon(iconId)
                .create()
                .show();

    }

    public void showAlertDialog(Context context, int titleId, int messageId,
                                int iconId,
                                int positiveTextId, DialogInterface.OnClickListener positiveButtonListener,
                                int negativeTextId, DialogInterface.OnClickListener negativeButtonListener) {

        new AlertDialog.Builder(context)
                .setTitle(titleId)
                .setMessage(messageId)
                .setPositiveButton(positiveTextId, positiveButtonListener)
                .setNegativeButton(negativeTextId, negativeButtonListener)
                .setIcon(iconId)
                .create()
                .show();

    }

    @Override
    protected void onResume() {
        super.onResume();
        ActivityManager.setCurrent(this);
    }

    @Override
    protected void onPause() {
        super.onPause();
    }
}

File: app/src/main/java/com/github/pwcong/jpstart/view/activity/BaseActivity.kt
package com.github.pwcong.jpstart.ui.activity

import android.content.Context
import android.content.DialogInterface
import android.os.Bundle

import android.view.View
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.viewbinding.ViewBinding

import com.google.android.material.snackbar.Snackbar

import com.github.pwcong.jpstart.manager.ActivityManager

abstract class BaseActivity<T : ViewBinding> : AppCompatActivity() {
    lateinit var viewBinding: T

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        ActivityManager.getInstance().register(this)
        ActivityManager.getInstance().setCurrent(this)

        viewBinding = initViewBinding()
        setContentView(viewBinding.root)

        init(savedInstanceState)
    }

    protected abstract fun initViewBinding(): T

    protected abstract fun init(savedInstanceState: Bundle?)

    override fun onDestroy() {
        super.onDestroy()

        ActivityManager.getInstance().unregister(this)
    }

    fun showSnackBar(view: View, msg: String) {
        Snackbar.make(view, msg, Snackbar.LENGTH_SHORT).show()
    }

    fun showSnackBar(view: View, msg: Int) {
        Snackbar.make(view, msg, Snackbar.LENGTH_SHORT).show()
    }

    fun showAlertDialog(
        context: Context, titleId: Int, message: String,
        iconId: Int,
        positiveTextId: Int, positiveButtonListener: DialogInterface.OnClickListener?
    ) {
        AlertDialog.Builder(context)
            .setTitle(titleId)
            .setMessage(message)
            .setPositiveButton(positiveTextId, positiveButtonListener)
            .setIcon(iconId)
            .create()
            .show()
    }

    fun showAlertDialog(
        context: Context, titleId: Int, messageId: Int,
        iconId: Int,
        positiveTextId: Int, positiveButtonListener: DialogInterface.OnClickListener?
    ) {
        AlertDialog.Builder(context)
            .setTitle(titleId)
            .setMessage(messageId)
            .setPositiveButton(positiveTextId, positiveButtonListener)
            .setIcon(iconId)
            .create()
            .show()
    }

    fun showAlertDialog(
        context: Context, titleId: Int, message: String,
        iconId: Int,
        positiveTextId: Int, positiveButtonListener: DialogInterface.OnClickListener?,
        negativeTextId: Int, negativeButtonListener: DialogInterface.OnClickListener?
    ) {
        AlertDialog.Builder(context)
            .setTitle(titleId)
            .setMessage(message)
            .setPositiveButton(positiveTextId, positiveButtonListener)
            .setNegativeButton(negativeTextId, negativeButtonListener)
            .setIcon(iconId)
            .create()
            .show()
    }

    fun showAlertDialog(
        context: Context, titleId: Int, messageId: Int,
        iconId: Int,
        positiveTextId: Int, positiveButtonListener: DialogInterface.OnClickListener?,
        negativeTextId: Int, negativeButtonListener: DialogInterface.OnClickListener?
    ) {
        AlertDialog.Builder(context)
            .setTitle(titleId)
            .setMessage(messageId)
            .setPositiveButton(positiveTextId, positiveButtonListener)
            .setNegativeButton(negativeTextId, negativeButtonListener)
            .setIcon(iconId)
            .create()
            .show()
    }

    override fun onResume() {
        super.onResume()
        ActivityManager.getInstance().setCurrent(this)
    }
}